File: cpp/32_pandigital_products.cpp
#include <iostream>
#include <string>
#include <set>

bool is_pandigital_product(int a, int b){
    std::string identity = std::to_string(a) + std::to_string(b) + std::to_string(a * b);
    std::set<char> seen(identity.begin(), identity.end());
    // result must be 1-9 pandigital, cannot contain 0
    return (seen.find('0') == seen.end() and seen.size() == identity.size());
}

int pandigital_product(){
    std::set<int> products;
    // 1-digit number * 4-digit number = 4-digit number
    // numbers 1000-1233, 9877-9999 contain either 0 or duplicate digits
    for (int i = 1; i <= 9; i++){
        for (int j = 1234; j <= 9876; j++){
            if (is_pandigital_product(i, j)){
                products.insert(i * j);
            }
        }
    }
    // 2-digit number * 3-digit number = 4-digit number
    // 10-11, 99, 100-122, 988-999 contain either 0 or duplicate digits
    for (int i = 12; i <= 98; i++){
        for (int j = 123; j <= 987; j++){
            if (is_pandigital_product(i, j)){
                products.insert(i * j);
            }
        }
    }
    int total = 0;
    for (int product : products){
        total += product;
    }
    return total;
}

int main(){
    std::cout << pandigital_product() << "\n";
}

File: c#/_32_PandigitalProducts.cs
using System;
using System.Collections.Generic;

public class _32_PandigitalProducts{
	public static int Euler32(){
		HashSet<int> products = new HashSet<int>();
		// 2-digit * 3-digit = 4-digit
		for (int a = 12; a <= 98; a++){
			for (int b = 123; b <= 987; b++){
				if (b % 10 == 0) continue;
				int product = a * b;
				if (IsPandigital(a, b, product)){
					products.Add(product);
				}
			}
		}
		// 1-digit * 4-digit = 4-digit
		for (int a = 2; a < 10; a++){
			for (int b = 1234; b <= 9876; b++){
				if (b % 10 == 0) continue;
				int product = a * b;
				if (IsPandigital(a, b, product)){
					products.Add(product);
				}
			}
		}
		int res = 0;
		foreach (int product in products){
			res += product;
		}
		return res;
	}
	
	public static bool IsPandigital(int a, int b, int c){
		string digits = a.ToString() + b.ToString() + c.ToString();
		if (digits.Length != 9) return false;
		bool[] seen = new bool[9];
		foreach (char digit in digits){
			int idx = digit - '1';
			if (idx < 0 || seen[idx]) return false;
			seen[idx] = true;
		}
		return true;
	}
	
	public static void Main(){
		Console.WriteLine(Euler32());
	}
}
